program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+use {
+    i64 as SignedTrackerRunningSum,
+    u64 as UnsignedTrackerRunningSum,
+};
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+pub trait Tracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    ///initializes a zero initialized tracker
+    fn initialize(&mut self) -> Result<(), OracleError>;
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError>;
+    ///gets the index of the next entry
+    fn get_next_entry(current_entry: usize) -> usize {
+        (current_entry + 1) % NUM_ENTRIES
+    }
+    ///Given a time, return the entry of that time
+    fn time_to_entry(time: i64) -> Result<usize, OracleError> {
+        Ok(try_convert::<i64, usize>(time / GRANUALITY)? % NUM_ENTRIES)
+    }
+    ///returns the remining time before the next multiple of GRANUALITY
+    fn get_time_to_entry_end(time: i64) -> i64 {
+        (GRANUALITY - (time % GRANUALITY)) % GRANUALITY
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an SMA Tracker that has NUM_ENTRIES entries
+/// each tracking time weighted sums for GRANUALITY seconds periods.
+///The prices are assumed to be provided under some fixed point representation, and the computation
+/// gurantees accuracy up to the last decimal digit in the fixed point representation.
+/// Assumes THRESHOLD < GRANUALITY
+pub struct SmaTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    valid_entry_counter:        u64, // is incremented everytime we add a valid entry
+    max_update_time_gap:        i64, // maximum time between two updates in the current entry.
+    running_price:              [SignedTrackerRunningSum; NUM_ENTRIES], /* price running time
+                                      * weighted sums */
+    running_confidence:         [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                                           * time
+                                                                           * weighted sums */
+    max_observed_entry_abs_val: u64, /* used to ensure that there are no overflows when
+                                      * looking at running sum diffs */
```

Comment:
let's say we're in a situation where this abs_val gets set high enough to break the SMA. If that happens, then we need to manually reset the state of the whole tracker to fix it?

Reply:
Yep, this isn't great. Let's discuss it in person

Reply:
I agree that this is not ideal, I believe that the best way is to have a type bigger than 64 bits, and that is big enough to support our granualirities. Otherwise, there's really not much you can do beyond invalidate an entry if an overflow happens with in it. The main risk created by this is that one very bad price is enough to ruin things forever.

I can have an alternative, where instead we store the slot of the last overflow, and use that to ensure that all intermediate overflow points are included in our Wormhole sequence. Retrospectively, this seems cleaner what do you think?

Alternatively, we can choose not to support longer time averages if we believe people are unlikely to use them.